### ISSUE TITLE
DDP (Multi-Node) inference error in utils/__init__.py

### ISSUE THREAD

**Issue by SimoneCaldarella** [ultralytics/ultralytics#20283]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Multi-GPU

### Bug

Hi, I got this error when instantiating YOLOv8 on a distributed scenario (device_id invalid ordinal) and found out that in `ultralytics/utils/__init__.py`, when the distributed context manager is initialized there is the following call
```
with torch_distributed_zero_first(RANK):
    ...
```
causing error in multi-node setting where the correct one should be
```
with torch_distributed_zero_first(LOCAL_RANK):
```

### Environment

Ultralytics 8.3.110 🚀 Python-3.11.7 torch-2.3.0+cu121 CPU (Intel Xeon Platinum 8358 2.60GHz)
Setup complete ✅ (128 CPUs, 503.0 GB RAM, 0.4/4.7 GB disk)

OS                  Linux-4.18.0-425.19.2.el8_7.x86_64-x86_64-with-glibc2.28
Environment         Linux
Python              3.11.7
Install             pip
Path                /leonardo/home/userexternal/scaldare/miniconda3/envs/dpo/lib/python3.11/site-packages/ultralytics
RAM                 502.99 GB
Disk                0.4/4.7 GB
CPU                 Intel Xeon Platinum 8358 2.60GHz
CPU count           128
GPU                 None
GPU count           None
CUDA                None

numpy               ✅ 1.26.4<=2.1.1,>=1.23.0
matplotlib          ✅ 3.9.0>=3.3.0
opencv-python       ✅ 4.9.0.80>=4.6.0
pillow              ✅ 11.1.0>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.13.1>=1.4.1
torch               ✅ 2.3.0>=1.8.0
torch               ✅ 2.3.0!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.18.0>=0.9.0
tqdm                ✅ 4.67.1>=4.64.0
psutil              ✅ 6.1.1
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.3>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.14>=2.0.0

### Minimal Reproducible Example

![Image](https://github.com/user-attachments/assets/ada8c720-3a38-4373-8733-1d953f774274)

### Additional

I am working on a large cluster hence the specifics are related to the login node. However, each cluster node contains 4 A100 64GB custom and 512GB of RAM.

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @SimoneCaldarella, thank you for your detailed report and for your willingness to contribute to Ultralytics! 🚀 This is an automated response to help you get started—an Ultralytics engineer will also assist you here soon.

We recommend checking out our [Docs](https://docs.ultralytics.com/) if you haven’t already, where you’ll find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) guides and answers to common questions.

Since this is a 🐛 Bug Report, please ensure you’ve provided a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE). If the image you attached is part of the MRE, that's helpful! If you can, please provide a concise code snippet or script that reliably reproduces the issue you’re seeing in your multi-node distributed environment.

To make sure the problem persists in the latest version, please upgrade your `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

### Environments

YOLO runs in several verified environments with all dependencies (including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/)) preinstalled:

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM: [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI: [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**: [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

### Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

### Community & Support

For real-time chat, join us on [Discord](https://discord.com/invite/ultralytics) 🎧. For in-depth discussions, check out [Discourse](https://community.ultralytics.com/). You can also dive into threads on our [Subreddit](https://reddit.com/r/Ultralytics) to connect with the community.

Thank you again for reporting this and offering to submit a PR! An Ultralytics engineer will review your suggestion and assist further soon.

**Comment by glenn-jocher** [comment]
Hi @SimoneCaldarella,

Thank you for reporting this issue with such a detailed explanation. You're absolutely right - using `RANK` instead of `LOCAL_RANK` in the distributed context manager would cause problems in multi-node settings.

The `torch_distributed_zero_first` context manager is designed to make processes wait for the rank 0 process to complete a task first. In multi-node setups, we should indeed be using `LOCAL_RANK` to ensure proper synchronization within each node.

Your proposed fix looks correct. We'd greatly appreciate your PR contribution to address this issue. Thank you for your willingness to help improve the codebase!

**Comment by SimoneCaldarella** [comment]
Hi @glenn-jocher,

I submitted the PR as described in the contribution documentation. I also commented to sign the CLA.
Bests!

**Comment by glenn-jocher** [comment]
Thank you @SimoneCaldarella for submitting the PR so quickly! I appreciate you following the contribution process and signing the CLA. We'll review your changes as soon as possible and get this fix merged to ensure proper distributed processing in multi-node environments. This improvement will be valuable for users working with YOLO models on large clusters like yours.

**Comment by Viki-researcher** [comment]
> Thank you [@SimoneCaldarella](https://github.com/SimoneCaldarella) for submitting the PR so quickly! I appreciate you following the contribution process and signing the CLA. We'll review your changes as soon as possible and get this fix merged to ensure proper distributed processing in multi-node environments. This improvement will be valuable for users working with YOLO models on large clusters like yours.

is multi-node train supported now?

**Comment by glenn-jocher** [comment]
Hi @Viki-researcher,

Yes, multi-node training is supported in Ultralytics. The PR from @SimoneCaldarella addresses a specific bug in the distributed context manager that affected multi-node setups. With this fix, the distributed training process will properly synchronize across nodes using `LOCAL_RANK` instead of `RANK`. This is particularly important for large-scale training on clusters with multiple machines, each containing multiple GPUs.
